src/mo_frontend/typing.ml

```diff
@@ -2637,9 +2646,13 @@ and infer_dec_valdecs env dec : Scope.t =
     let obj_typ = object_of_scope env obj_sort.it dec_fields obj_scope' at in
     let _ve = check_pat env obj_typ pat in
     Scope.{empty with val_env = T.Env.singleton id.it obj_typ}
-  | LetD (pat, exp) ->
+  | LetD (pat, exp, fail) ->
     let t = infer_exp {env with pre = true} exp in
-    let ve' = check_pat_exhaustive (if is_import dec then local_error else warn) env t pat in
+    let consequence = match fail with
+      | None when is_import dec -> local_error
+      | None -> warn
+      | Some _ -> fun _ _ _ -> Format.kasprintf ignore in
+    let ve' = check_pat_exhaustive consequence env t pat in
```

Comment:
I would maybe split this into two cases, one for Let(pat, exp, None) and one for Let(pat, exp, Some fail).

In the second case, I'd call check_pat instread of check_pat_exhaustive with an error function that discard the error.

Something like

```
| LetD (pat, exp, None) ->
    let t = infer_exp {env with pre = true} exp in
    let warnOrError =
      if is_import dec then
        local_error
      else warn
    in
    let ve' = check_pat_exhaustive warnOrError env t pat in
    Scope.{empty with val_env = ve'}
  | LetD (pat, exp, Some fail) ->
    let t = infer_exp {env with pre = true} exp in
    let ve' = check_pat env t pat in (* need not be exaustive *)
    Scope.{empty with val_env = ve'}
```

Or maybe even just:

```
| LetD (pat, exp, fail) ->
    let t = infer_exp {env with pre = true} exp in
    let ve' = match fail with
      | None ->
        check_pat_exhaustive
         (if is_import dec then local_error else warn)
         env t pat
      | Some _exp1 ->
         check_pat env t pat
    in
    Scope.{empty with val_env = ve'}

```